capsules/src/hmac.rs

```diff
@@ -0,0 +1,488 @@
+                    }
+                })
+        });
+    }
+
+    fn hash_done(&'a self, result: Result<(), ReturnCode>, digest: &mut T) {
+        self.appid.map(|id| {
+            self.apps
+                .enter(*id, |app, _| {
+                    self.hmac.clear_data();
+
+                    let pointer = digest.as_ref()[0] as *mut u8;
+
+                    match app.dest.as_mut() {
+                        Some(dest) => {
+                            dest.as_mut().copy_from_slice(digest.as_ref());
+                        }
+                        None => {}
+                    };
+
+                    app.callback.map(|cb| match result {
+                        Ok(_) => cb.schedule(0, pointer as usize, 0),
+                        Err(e) => cb.schedule(usize::from(e), pointer as usize, 0),
+                    })
+                })
+                .map_err(|err| {
+                    if err == kernel::procs::Error::NoSuchApp
+                        || err == kernel::procs::Error::InactiveApp
+                    {
+                        self.appid.clear();
+                    }
+                })
+        });
+    }
+}
+
+/// Specify memory regions to be used.
+///
+/// ### `allow_num`
+///
+/// - `0`: Allow a buffer for storing the key.
+///        The kernel will read from this when running
+///        This should not be changed after running `run` until the HMAC
+///        has completed
+/// - `1`: Allow a buffer for storing the buffer.
+///        The kernel will read from this when running
+///        This should not be changed after running `run` until the HMAC
+///        has completed
+/// - `2`: Allow a buffer for storing the digest.
+///        The kernel will fill this with the HMAC digest before calling
+///        the `hash_done` callback.
+impl<H: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> Driver for HmacDriver<'a, H, T> {
+    fn allow(
+        &self,
+        appid: AppId,
+        allow_num: usize,
+        slice: Option<AppSlice<Shared, u8>>,
+    ) -> ReturnCode {
+        // Return true if this app already owns the HMAC capsule, if no app owns
+        // the HMAC capsule, or if the app that is marked as owning the HMAC
+        // capsule no longer exists.
+        let match_or_empty_or_nonexistant = self.appid.map_or(true, |owning_app| {
+            if self.active.get() {
+                owning_app == &appid
+            } else {
+                self.apps
+                    .enter(*owning_app, |_, _| owning_app == &appid)
+                    .unwrap_or(true)
+            }
+        });
+        if match_or_empty_or_nonexistant {
+            self.appid.set(appid);
+        } else {
+            return ReturnCode::ENOMEM;
+        }
+        match allow_num {
+            // Pass buffer for the key to be in
+            0 => self.appid.map_or(ReturnCode::FAIL, |id| {
+                self.apps
+                    .enter(*id, |app, _| {
```

Comment:
I think because we have a grant we can remove all of the checks for driver ownership (i.e. `self.appid`) for allow and subscribe. Instead, the buffers and callback can just be unconditionally stored.